docs/spec/annotations.rst

```diff
@@ -53,43 +53,154 @@ decorators ``@property``, ``@staticmethod`` and ``@classmethod``.
+                   : | `tuple_type_expression`
+                   : | <Annotated> '[' `type_expression` ',' expression (',' expression)* ']'
+                   : | <TypeGuard> '[' `type_expression` ']'
+                   :       (valid only in some contexts)
+                   : | <TypeIs> '[' `type_expression` ']'
+                   :       (valid only in some contexts)
+                   : | `string_annotation`
+                   :       (must evaluate to a valid `type_expression`)
+    maybe_unpacked: `type_expression` | `unpacked`
+    unpacked: '*' `unpackable``
+            : | <Unpack> '[' `unpackable` ']'
+    unpackable: `tuple_type_expression``
+              : | name
+              :       (where name must refer to an in-scope TypeVarTuple)
+    tuple_type_expression: <tuple> '[' '(' ')' ']'
+                         :      (representing an empty tuple)
+                         : | <tuple> '[' `type_expression` ',' '...' ']'
+                         :       (representing an arbitrary-length tuple)
+                         : | <tuple> '[' `maybe_unpacked` (',' `maybe_unpacked`)* ']'
+    string_annotation: string
+                     :     (must be a string literal that is parsable as Python code; see "String annotations")
+    type_expression_list: '[' `type_expression` (',' `type_expression`)* ']'
+                        : '[' ']'
+
+Notes:
+
+* The grammar assumes the code has already been parsed as Python code, and
+  loosely follows the structure of the AST. Syntactic details like comments
+  and whitespace are ignored.
+
+* ``<Name>`` refers to a :term:`special form`. Most special forms must be imported
+  from :py:mod:`typing` or ``typing_extensions``, except for ``None``, ``type``,
+  ``tuple``, and ``InitVar``. The latter two have aliases in :py:mod:`typing`: :py:class:`typing.Type`
+  and :py:class:`typing.Tuple`.  ``InitVar`` must be imported from :py:mod:`dataclasses`.
+  ``Callable`` may be imported from either :py:mod:`typing` or :py:mod:`collections.abc`.
+  Special forms may be aliased
+  (e.g., ``from typing import Literal as L``), and they may be referred to by a
+  qualified name (e.g., ``typing.Literal``). There are other special forms that are not
+  acceptable in any annotation or type expression, including ``Generic``, ``Protocol``,
+  and ``TypedDict``.
+
+* Any leaf denoted as ``name`` may also be a qualified name (i.e., ``module '.' name``).
+
+* Comments in parentheses denote additional restrictions not expressed in the
+  grammar, or brief descriptions of the meaning of a construct.
+
+.. _ `string-annotations`:
 
 .. _`forward-references`:
 
-Forward references
+String annotations
 ------------------
 
-When a type hint contains names that have not been defined yet, that
+When a type hint cannot be evaluated at runtime, that
 definition may be expressed as a string literal, to be resolved later.
+Such annotations must be plain string literals, not raw strings, byte
+strings or f-strings.
```

Comment:
```suggestion
```

Saving this for https://discuss.python.org/t/exotic-kinds-of-string-annotations/50527